### PR TITLE
Support for FSDP + Ulysses and FSDPv2 + TP + SP Parallelism

### PR DESCRIPTION
**Pull request by lausannel**
This PR introduces support for two type of combination of parallelism and a feature for initialization:

- [X]  FSDPv2 + Ulysses
- [X] FSDPv2 + TP + SP
- [X] Low cpu memory usage for ulysses parallelism initialization
- [X] Llama3 model support for TP + SP
- [X] Align the training accuracy of ulysses parallel training and single GPU training
- [x] Align the training accuracy of tensor parallel training and single GPU training

To run the combinations of parallelism, use the following commands:

```bash
# fsdpv2 + ulysses
CUDA_VISIBLE_DEVICES=0,1,2,3 PJRT_ALLOCATOR_FRACTION=0.92 XLA_FLAGS="--xla_disable_hlo_passes=gpu-convert-async-collectives-to-sync,triton-autotuner" \
XLA_USE_SPMD=1  ./examples/run.sh --model ./hf_models/config/llama-3-1b --mbs 1 --fsdp 4 --sp_num 2 --spmd_fsdp --max_seq_length 8192

# fsdpv2 + tp + sp
CUDA_VISIBLE_DEVICES=0,1,2,3 PJRT_ALLOCATOR_FRACTION=0.92 XLA_FLAGS=" --xla_disable_hlo_passes=gpu-convert-async-collectives-to-sync,triton-autotuner" \
XLA_USE_SPMD=1 ./examples/run.sh --model ./hf_models/config/llama-3-1b --mbs 1 --fsdp 2 --tp 2 --sp --spmd_fsdp --max_seq_length 8192
```

### CODE REVIEW

File: flashmodels/datasets/hf_dataset.py [hunk 1]
```diff
@@ -65,16 +66,19 @@ def group_texts(examples):
...
     bs = args.micro_batch_size
     if args.tp_num > 1:
         bs *= args.dp_num
     if args.pp_num > 1:
         bs *= args.gradient_accumulation_steps
     if args.spmd_fsdp:
-        bs *= args.fsdp_num
+        dp_num = args.fsdp_num // args.sp_num
+        if dp_num != 0:
+            bs *= args.fsdp_num // args.sp_num
```
Comment (baoleai) [inline flashmodels/datasets/hf_dataset.py]: bs *= dp_num

File: flashmodels/patch/llama_model.py [hunk 2]
```diff
@@ -1,24 +1,283 @@
...
+        past_key_values: Cache,
+        output_attentions: bool,
+    ):
+        # TODO: As of torch==2.2.0, the `attention_mask` passed to the model in `generate` is 2D and of dynamic length even when the static
+        # KV cache is used. This is an issue for torch.compile which then recaptures cudagraphs at each decode steps due to the dynamic shapes.
+        # (`recording cudagraph tree for symint key 13`, etc.), which is VERY slow. A workaround is `@torch.compiler.disable`, but this prevents using
+        # `fullgraph=True`. See more context in https://github.com/huggingface/transformers/pull/29114
+
+        if self.config._attn_implementation == "flash_attention_2":
+            if attention_mask is not None and 0.0 in attention_mask:
```
Comment (baoleai) [inline flashmodels/patch/llama_model.py]: 0.0 in attention_mask will cause tensor evaluation in XLA, so here I suggest to replace the whole function as follows:
```
if attention_mask is not None:
    return attention_mask
return None
```

File: flashmodels/accelerators/acc_llama_accelerator.py [hunk 3]
```diff
@@ -364,12 +379,16 @@ def _forward_ag_sp(m, *args, **kwargs):
...
             return output
 
-        row_dim = 0 if self.args.use_zero3 else None
-        col_dim = 1 if self.args.use_zero3 else None
+        row_dim = 0 if (self.args.use_zero3
+                        or self.args.fsdp_num > 1) else None
+        col_dim = 1 if self.args.use_zero3 or self.args.fsdp_num > 1 else None
+
         device = lazy_device()
+        gc_cnt = self.args.gc_cnt
```
Comment (baoleai) [inline flashmodels/accelerators/acc_llama_accelerator.py]: unused?

**Review by baoleai — APPROVED** [review]

File: flashmodels/accelerators/acc_llama_accelerator.py (hunk 3 above)
Comment (lausannel) [reply flashmodels/accelerators/acc_llama_accelerator.py]: fixed

File: flashmodels/patch/llama_model.py (hunk 2 above)
Comment (lausannel) [reply flashmodels/patch/llama_model.py]: fixed

File: flashmodels/datasets/hf_dataset.py (hunk 1 above)
Comment (lausannel) [reply flashmodels/datasets/hf_dataset.py]: fixed
